### ISSUE TITLE
run sample.py will met an error when set honor_grid=True

### ISSUE THREAD

**Issue by xichaoqiang** [keurfonluu/fteikpy#21]
Hi,Sir 
I try run the sample.py by using marmousi velocity model.

But ,when set honor_grid=True. 

Runtime Error will appears.

![OQU7O{ YIF`$%2T29J8E847](https://user-images.githubusercontent.com/30900789/184082228-b939dd75-5d6d-40a9-b916-2a1cecc1ffa9.png)

**Comment by xichaoqiang** [comment]
Thanks you very much.

**Comment by keurfonluu** [comment]
Hi @xichaoqiang,

Can you try the development branch and see if it worked?

```
pip install -U https://github.com/keurfonluu/fteikpy/archive/devel.zip
```

I am not sure if it's fixed, but it seems to work on my end.

Thanks!

**Comment by xichaoqiang** [comment]
Thanks for your kindly reply. The runtime error is disappear. It seems will lost some ray points.

**Comment by xichaoqiang** [comment]
![image](https://user-images.githubusercontent.com/30900789/184459738-cee291c5-f08b-4f00-a57c-99566333b825.png)

**Comment by keurfonluu** [comment]
I don't see anything wrong, here. If you look at the X coordinates, from 4.24 to 4.3, the ray path is stopping on grid points, which indicates that the ray is travelling rather horizontally.

**Comment by keurfonluu** [comment]
The issue wasn't actually fixed but it should be now. Please update to version 2.3.4: `pip install fteikpy -U`.

**Comment by xichaoqiang** [comment]
![image](https://user-images.githubusercontent.com/30900789/184475657-6f7a45f1-5198-48c9-9b65-1a0a2fe5cd09.png)

**Comment by xichaoqiang** [comment]
I plot the ray path with x from 4.24 to 4.29.

You an see the ray path cross the grid with two points.

We only get one from the program.

**Comment by xichaoqiang** [comment]
let me try 2.3.4

**Comment by xichaoqiang** [comment]
I try modify the shrink function , the result seems better.

_ray2d.py

```
        if honor_grid:
            fac = shrink(pcur, delta, lower, upper)
            pcur -= fac * delta
            pcur[0] = min(max(pcur[0], z[0]), z[-1])
            pcur[1] = min(max(pcur[1], x[0]), x[-1])

            # if fac < 1.0:
            #     # Grid magnetism: handle precision issues due to fac
            #     for ix in range(2):
            #         if np.abs(pcur[ix] - lower[ix]) < 1.0e-8:
            #             pcur[ix] = lower[ix]

            #         elif np.abs(pcur[ix] - upper[ix]) < 1.0e-8:
            #             pcur[ix] = upper[ix]

            i = np.searchsorted(z, pcur[0], side="right") - 1
            j = np.searchsorted(x, pcur[1], side="right") - 1
            lower[0] = z[max(i - 1, 0)] if pcur[0] == z[i] else z[i]
            lower[1] = x[max(j - 1, 0)] if pcur[1] == x[j] else x[j]
            upper[0] = z[i + 1]
            upper[1] = x[j + 1]

            ray[count] = pcur.copy()
            count += 1
```

_common.py
```
@jitted("f8(f8[:], f8[:], f8[:], f8[:])")
def shrink(pcur, delta, lower, upper):
    """Stepsize shrinking factor if ray crosses interface."""
    lu=np.vstack((lower,upper))
    del_s=delta!=0
    delta[delta==0]=np.finfo(np.float32).eps
    sd=(np.sign(-delta)*0.5+0.5).astype(np.int_)
    if len(lower==2):
        nb=np.array((lu[sd[0],0],lu[sd[1],1]))
    elif len(lower==3):
        nb=np.array((lu[sd[0],0],lu[sd[1],1],lu[sd[2],2]))
    return np.min((pcur[del_s]-nb[del_s] )/delta[del_s] )
```

**Comment by xichaoqiang** [comment]
![image](https://user-images.githubusercontent.com/30900789/184478101-48e5e474-1f6f-464e-a5da-180036582a32.png)

**Comment by keurfonluu** [comment]
It may work for this case, but it does not pass the integration tests. Anyway, the shrinking method is smart, but it causes too much trouble for the little performance gain it brings compared to a brute force voxel traversal algorithm. I think I'll just discard it and adapt Amanatides and Woo's algorithm (see http://www.cse.yorku.ca/~amana/research/grid.pdf).

**Comment by xichaoqiang** [comment]
Cool. Thanks for your kindly reply. :)

**Comment by keurfonluu** [comment]
Actually found a small bug in `shrink` with a missing condition. Can you try the development version and see if it works?

```
pip install -U https://github.com/keurfonluu/fteikpy/archive/devel.zip
```

Thanks!

**Comment by xichaoqiang** [comment]
This new version seems give awesome results.
Very great.

**Comment by keurfonluu** [comment]
> This new version seems give awesome results. Very great.

The fix for this issue is now available in version 2.4.0.
